src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs

```diff
@@ -0,0 +1,348 @@
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(symbol.Kind);
+            }
+
+            bool isNamedTypeAccessibleWithin(INamedTypeSymbol type, ISymbol within0)
+            {
+                Debug.Assert(within0 is INamedTypeSymbol || within0 is IAssemblySymbol);
+                Debug.Assert(type != null);
+
+                if (!type.IsDefinition)
+                {
+                    foreach (var typeArg in type.TypeArguments)
+                    {
+                        if (!IsAccessibleWithin(typeArg, within0))
+                        {
+                            return false;
+                        }
+                    }
+                }
+
+                var containingType = type.ContainingType;
+                if (containingType == null)
+                {
+                    return isNonNestedTypeAccessible(type.ContainingAssembly, type.DeclaredAccessibility, within0);
+                }
+                else
+                {
+                    return isMemberAccessible(containingType, type.DeclaredAccessibility, within0, throughTypeOpt0: null);
+                }
+            }
+
+            bool isNonNestedTypeAccessible(IAssemblySymbol assembly, Accessibility declaredAccessibility, ISymbol within0)
+            {
+                Debug.Assert(within0 is INamedTypeSymbol || within0 is IAssemblySymbol);
+                Debug.Assert(assembly != null);
+
+                switch (declaredAccessibility)
+                {
+                    case Accessibility.Public:
+                        // Public symbols are always accessible from any context
+                        return true;
+
+                    case Accessibility.Private:
+                    case Accessibility.Protected:
+                    case Accessibility.ProtectedAndInternal:
+                        // Shouldn't happen except in error cases as these access levels cannot be used to declare top-level types.
+                        return false;
+
+                    case Accessibility.Internal:
+                    case Accessibility.ProtectedOrInternal:
+                        // An internal type is accessible if we're in the same assembly or we have
+                        // friend access to the assembly it was defined in.
+                        var withinAssembly = (within0 as IAssemblySymbol) ?? ((INamedTypeSymbol)within0).ContainingAssembly;
+                        return hasInternalAccessTo(withinAssembly, assembly);
+
+                    case Accessibility.NotApplicable:
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(declaredAccessibility);
+                }
+            }
+
+            bool hasInternalAccessTo(IAssemblySymbol fromAssembly, IAssemblySymbol toAssembly)
+            {
+                if (fromAssembly.Identity.Equals(toAssembly.Identity))
```

Comment:
>fromAssembly.Identity.Equals(toAssembly.Identity) [](start = 20, length = 49)

I don't think this condition is appropriate to check for containment, however it appears that this is what it is used for.

Reply:
This is being used to determine if the two assembly symbols identify the same assembly.

Reply:
It is not being used for containment *per se*. It is used in the sense that if a program has both symbols loaded, then they are part of the same assembly and can access internals.

---
In reply to: [185161623](https://github.com/dotnet/roslyn/pull/26367#discussion_r185161623) [](ancestors = 185161623)

Reply:
> It is not being used for containment per se. It is used in the sense that if a program has both symbols loaded, then they are part of the same assembly and can access internals.

I argue that this logic is not correct. Equality of identities  says nothing about whether the assembly is the same.

---
In reply to: [185162836](https://github.com/dotnet/roslyn/pull/26367#discussion_r185162836) [](ancestors = 185162836,185161623)